library/lua/gui/dwarfmode.lua

```diff
@@ -436,21 +438,51 @@ MenuOverlay = defclass(MenuOverlay, DwarfOverlay)
 MenuOverlay.ATTRS {
     frame_inset = 0,
     frame_background = gui.CLEAR_PEN,
+
+    -- if sidebar_mode is set, we will enter the specified sidebar mode on show
+    -- and restore the previous sidebar mode on dismiss. otherwise it is up to
+    -- the caller to ensure we are in a sidebar mode where the menu is visible.
+    sidebar_mode = DEFAULT_NIL,
 }
 
 function MenuOverlay:computeFrame(parent_rect)
     return self.df_layout.menu, gui.inset_frame(self.df_layout.menu, self.frame_inset)
 end
 
-function MenuOverlay:onAboutToShow(below)
-    MenuOverlay.super.onAboutToShow(self,below)
+function MenuOverlay:onAboutToShow(parent)
+    if not dfhack.isMapLoaded() then
+        qerror('Please load a fortress map.')
+    end
+
+    self.saved_sidebar_mode = df.global.ui.main.mode
+    -- if we can't get back to the saved sidebar mode via enterSidebarMode
+    -- then fall back to the default mode
+    if dfhack.gui.getCurFocus(true):find('^dfhack/')
```

Comment:
Looking at `enterSidebarMode()`, it doesn't appear that it cares about `^dfhack/` as long as it can Escape out of the mode a few times and get back to `dwarfmode/Default`. Is there a particular case you're handling here? Should it be documented, or perhaps this test should be factored out into a `canEnterSidebar()` test function?

Reply:
I'll add a comment. the case we're handling here is that if a `dfhack` window is on the screen, then we can't trust that the current sidebar mode was set by the user. it could be a previous `MenuOverlay` subclass that is currently being shown, and it has set the sidebar mode for its own purposes.